### PR TITLE
Add normalization to sum-product, avoiding underflow.

### LINKED ISSUE #1448: Incorrect MPE for Long Chains of Discrete Factor Graphs
## Description
Hi team, thanks for the wonderful library!

I've run into an issue where I have a fairly large Discrete Factor Graph that I believe is running into an underflow type of error. I've copied a minimal toy example below that shows the same symptom. I'm not sure where to turn for documentation of python bindings or how to debug the optimization further. Thank you for the help!

## Steps to reproduce
Toy example is a hidden markov model with 3 possible states and 200 Observations. In this example, the transition matrix is mostly identity. The desired state outputs any number [0, 199] with equal probability. All other states output 200 with probability 1. I set each measurement equal to the observation index (so only [0,200] are observed and the desired state should be preferred always).

``` python
import gtsam
import numpy as np

import gtbook
from gtbook.discrete import Variables
VARIABLES = Variables()

def toy_problem(num_states, num_obs, desired_idx):
    states = list(range(num_states))
    X = VARIABLES.discrete_series("X", range(num_obs), states)
    Z = VARIABLES.discrete_series("Z", range(num_obs), range(num_obs+1))
    graph = gtsam.DiscreteFactorGraph()

    # Mostly identity transition matrix
    transitions = np.eye(num_states)
    
    # Needed otherwise mpe is always state 0?
    transitions += 0.1/(num_states)
    
    transition_cpt = []
    for i in range(0, num_states):
        transition_row = "/".join([str(x) for x in transitions[i]])
        transition_cpt.append(transition_row)
    transition_cpt = " ".join(transition_cpt)

    for i in reversed(range(1, num_obs)):
        transition_conditional = gtsam.DiscreteConditional(X[i], [X[i-1]], transition_cpt)
        graph.push_back(transition_conditional)

    # Contrived example such that the desired state gives measurements [0, num_obs) with equal probability
    #   but all other states always give measurement num_obs
    obs = np.zeros((num_states, num_obs+1))
    obs[:,-1] = 1
    obs[DESIRED_STATE,0: -1] = 1
    obs[DESIRED_STATE,-1] = 0
    obs_cpt_list = []
    for i in range(0, num_states):
        obs_row = "/".join([str(z) for z in obs[i]])
        obs_cpt_list.append(obs_row)
    obs_cpt = " ".join(obs_cpt_list)
    
    # Contrived example where each measurement is its own index
    for i in range(0, num_obs):
        obs_conditional = gtsam.DiscreteConditional(Z[i], [X[i]], obs_cpt)
        factor = obs_conditional.likelihood(i)
        graph.push_back(factor)
    
    mpe = graph.optimize()
    vals = [mpe[X[i][0]] for i in range(num_obs)]
    print("MPE Vals: ")
    print(vals)
    print("Indices where mpe does not equal desired state:")
    print(np.where(np.array(vals) != DESIRED_STATE))

    # Missing many outputs
    sum_product = graph.sumProduct()
    breakpoint()

if __name__ == "__main__":
    NUM_STATES = 3
    NUM_OBS = 200
    DESIRED_STATE = 1

    toy_problem(NUM_STATES, NUM_OBS, DESIRED_STATE)
```

Output:
```
MPE Vals: 
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
Indices where mpe does not equal desired state:
(array([138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150,
       151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163,
       164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176,
       177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189,
       190, 191, 192, 193, 194, 195, 196, 197, 198, 199]),)

```

## Expected behavior

Expected behavior is that the maximum probability estimate is equal to DESIRED_STATE (1) for all hidden states. This does not hold true for hidden states 138 and above. Additionally, the graph resulting from `sum_product = graph.sumProduct()` is missing many values for these incorrect states.

## Environment

I'm running on Python 3.8.13, gtsam version 4.2a9, Ubuntu 20.04

### PR DESCRIPTION
**Pull request by keevindoherty**
This quick fix should actually resolve #1448 now. Adds the same product normalization from #1449 to the `DiscreteFactorGraph::DiscreteEliminate` function. This broke an existing test in `testDiscreteFactorGraph.cpp` due to an expected result being unnormalized, where the result from `DiscreteEliminate` was normalized, so I modified the test to expect a normalized result. I also added a "visual test" to `test_DiscreteFactorGraph.py` to verify the expected results in #1448.

Couple notes for reviewer(s):
- *Highest priority:* The test I added in `test_DiscreteFactorGraph.py` is not a real test, i.e. it does not computationally verify the result and report back, just "prints some stuff." It should not be merged in its current state, but I don't think that simply verifying the number of assignments in a sum-product result is the right move here. If there is a suggestion for a more robust test, I'm happy to cook something up and modify accordingly, but I want to make sure we're on the same page before I try throwing something together.
- *Low priority:* The patch I added in `testDiscreteFactorGraph.cpp` normalizes the expect result before comparison, but it may be worth thinking longer term about mechanisms for normalizing discrete factor graph results (e.g. should discrete conditionals simply be normalized be default? or not, in which case one option would be to add a new assert for discrete to compare "up to scale" where appropriate?). This seems not urgent but perhaps worth earmarking. The latter seems particularly useful if we want to keep constructing "expected" results for future tests without normalizing "by hand." Part of my thinking on this is: I'm not actually sure personally how to implement the same patch in `testDiscreteFactorGraph.cpp` if we had to do it using the GTSAM Python API; is it possible? I guess the last possibility would be to add something like a `DecisionTreeFactor& DecisionTreeFactor::normalize() const` that returns a normalized copy of a DecisionTreeFactor?

Let me know what you think?

### CODE REVIEW

**Comment by dellaert** [comment]
> * _Highest priority:_ The test I added in `test_DiscreteFactorGraph.py` is not a real test,...If there is a suggestion for a more robust test, I'm happy to cook something up

So, I have been thinking of HMMs, as you know, and a single sweep of an HMM is very easy to code up iteratively, hence a good test would be to compare a long HMM with the last marginal result of elimination.

Alternatively, even easier is a measurement-free Markov chain: The marginal at the end should be T^n which will be the stationary distribution, i.e. Eigenvector of T.

> * _Low priority:_ The patch I added in `testDiscreteFactorGraph.cpp` normalizes the expect result before comparison, but it may be worth thinking longer term about mechanisms for normalizing discrete factor graph results

Discrete factors are not normalized (and can even be confsing if we do). Maybe "normalizing" by max would be better to not confuse peeople.

DiscreteConditionals should already be normalized.

**Comment by keevindoherty** [comment]
> Alternatively, even easier is a measurement-free Markov chain: The marginal at the end should be T^n which will be the stationary distribution, i.e. Eigenvector of T.

I took a first crack at implementing this test in the latest commit. Maybe you or Varun can take a look when you get a chance?

> DiscreteConditionals should already be normalized.

Ah, yep, sorry, I got my wires crossed here. The issue was that the comparison in `testDiscreteFactorGraph.cpp` was comparing two *DecisionTreeFactor* objects, which are _not_ normalized (and probably should not be?), so some extra care needs to be taken when constructing an "expected" DecisionTreeFactor to ensure normalization for testing discrete.

**Comment by dellaert** [comment]
> DecisionTreeFactor

No, DecisionTreeFactors are not normalized as a rule, unless they are the base class of a Conditional

File: gtsam/discrete/DiscreteFactorGraph.cpp [hunk 1]
```diff
@@ -210,6 +210,12 @@ namespace gtsam {
     for (auto&& factor : factors) product = (*factor) * product;
     gttoc(product);
 
+    // Sum all the potentials by pretending all keys are frontal:
+    auto normalization = product.sum(product.size());
```
Comment (dellaert) [inline gtsam/discrete/DiscreteFactorGraph.cpp:214]: Wondering whether we could use `max` here to avoid the impression the factor needs to be normalized…

**Review by dellaert — APPROVED** [review]
Awesome!!! One comment about max. I think sum creates a wrong expectation…

**Comment by keevindoherty** [comment]
Yeah, max is reasonable too. “Normalizing” w.r.t. the max seems to be the usual technique for dealing with this when performing computations in log-space, too (Koller and Friedman discuss it, if I remember correctly).

I also was wondering if there was a nicer way to extract the probabilities from a DiscreteConditional in the Python API than the “brute force” method I used (see test case)?

File: python/gtsam/tests/test_DiscreteFactorGraph.py [hunk 2]
```diff
@@ -216,5 +216,71 @@ def make_key(character, index, cardinality):
...
+        # Get the DiscreteConditional representing the marginal on the last factor
+        last_marginal = sum_product.at(chain_length - 1)
+
+        # Extract the actual marginal probabilities
+        assignment = DiscreteValues()
+        marginal_probs = []
+        for i in range(num_states):
+            assignment[X[chain_length - 1][0]] = i
+            marginal_probs.append(last_marginal(assignment))
+        marginal_probs = np.array(marginal_probs)
```
Comment (keevindoherty) [inline python/gtsam/tests/test_DiscreteFactorGraph.py:280]: I wasn't sure if there was a nicer way to extract these probabilities from the discrete conditional (i.e. simpler than 'brute force' iterating over each assignment)

**Comment by dellaert** [comment]
> Yeah, max is reasonable too. “Normalizing” w.r.t. the max seems to be the usual technique for dealing with this when performing computations in log-space, too (Koller and Friedman discuss it, if I remember correctly).
> 
> I also was wondering if there was a nicer way to extract the probabilities from a DiscreteConditional in the Python API than the “brute force” method I used (see test case)?

I would prefer to go with max: the existence of a 1.0 in the decision tree somewhere would be additional verification we are talking about a factor and not a conditional

File: python/gtsam/tests/test_DiscreteFactorGraph.py (hunk 2 above)
Comment (dellaert) [reply python/gtsam/tests/test_DiscreteFactorGraph.py:280]: Hmm, yes, there must be. Let me think...

Comment (dellaert) [reply python/gtsam/tests/test_DiscreteFactorGraph.py:280]: So, there does not seem to be a nice method for this, but a trick used in test_DiscreteConditional.py is to go the opposite way: create an *expected* marginal by specifying a vector of doubles, and then use gtsamAssertEquals:
```python
        expected = DecisionTreeFactor(Y, "0.8 0.6 0.5")
        self.gtsamAssertEquals(actual1, expected, 1e-9)
```

Comment (dellaert) [reply python/gtsam/tests/test_DiscreteFactorGraph.py:280]: Note you can also use a vector<double>

**Comment by keevindoherty** [comment]
> I would prefer to go with max: the existence of a 1.0 in the decision tree somewhere would be additional verification we are talking about a factor and not a conditional

Sounds good to me. I can make this change tomorrow or Monday so we can get this merged. Just to check - we’ll normalize to the max in EliminateForMPE (max product) too, right? And then is it OK with you if that’s on this same PR? Should just be a one line change, I think.

File: python/gtsam/tests/test_DiscreteFactorGraph.py (hunk 2 above)
Comment (keevindoherty) [reply python/gtsam/tests/test_DiscreteFactorGraph.py:280]: This looks good to me, and looks a bit cleaner, I’ll do that.

**Comment by dellaert** [comment]
> > I would prefer to go with max: the existence of a 1.0 in the decision tree somewhere would be additional verification we are talking about a factor and not a conditional
> 
> Sounds good to me. I can make this change tomorrow or Monday so we can get this merged. Just to check - we’ll normalize to the max in EliminateForMPE (max product) too, right? And then is it OK with you if that’s on this same PR? Should just be a one line change, I think.

Yes, absolutely. Note I don't know whether there already is a `max`, but should be easy enough, cribbing from sum.

File: python/gtsam/tests/test_DiscreteFactorGraph.py [hunk 3]
```diff
@@ -216,5 +216,63 @@ def make_key(character, index, cardinality):
...
+        transitions /= np.sum(transitions, axis=0)
+
+        # The stationary distribution is the eigenvector corresponding to eigenvalue 1
+        eigvals, eigvecs = np.linalg.eig(transitions)
+        stationary_idx = np.where(np.isclose(eigvals, 1.0))
+        stationary_dist = eigvecs[:, stationary_idx]
+
+        # Ensure that the stationary distribution is positive and normalized
+        stationary_dist /= np.sum(stationary_dist)
+        expected = DecisionTreeFactor(X[chain_length-1], stationary_dist.flatten())
```
Comment (dellaert) [inline python/gtsam/tests/test_DiscreteFactorGraph.py:253]: Awesome !!!!

**Comment by dellaert** [comment]
Killing all but required CI and will merge after required passes.

**Comment by dellaert** [comment]
Thanks Kevin, this is awesome !

**Pull request merged**